### ISSUE TITLE
Remove IE-specific code/libraries

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#214]
The current state of our codebase does have support for IE8.  Drupal 8 has dropped it.

Facts worth considering:
- The combined, global usage of IE versions 5 through 10 is less than ~0.4% (caniuse.com/usage-table). Actually all IE versions + all Edge versions seem to be around ~7% (w3counter.com/globalstats.php). ...although to be fair, it seems that usage is more around 10%-11% if you limit the scope to desktop browsers alone (netmarketshare.com/browser-market-share.aspx). Roughly speaking, everyone seems to be using either Chrome, or Safari these days (http://gs.statcounter.com).

- Google supports only IE11 and Edge (https://support.google.com/drive/answer/2375082?hl=en)

What do you think?

**Comment by serundeputy** [comment]
drop;

**Comment by MrHaroldA** [comment]
What if Backdrop core only supports "modern" browsers and leave support for IE8 (and IE9?) up to the contributed themes? Would that be such a problem?

I would even suggest that people building and/or administrating websites don't use IE8 anymore as they are a bit more techy than the average Internet user.

**Comment by wesruv** [comment]
We could do a "partial support". Say that IE8 gets the mobile version of the admin and core themes, but everything should be functional, likely much less beautiful, but no broken layout, forms work, etc. 

That's a much easier bar for us and gets the unfortunate souls supporting IE8 a good start. They'd likely just update CSS and likely just for the default theme, I doubt we're talking about too many clients that want to edit the site with IE8 at this point. 

FWIW, I have *not* been on a Lullabot or side project yet where I didn't have to support ie8 for the end users.

**Comment by klonos** [comment]
If we do decide to drop it, the we should at least consider including Modernizr [the library](http://modernizr.com/) / [the module](https://www.drupal.org/project/usage/modernizr) in order to give site developers that are still required to support the old browsers a hand in that task.

**Comment by MrHaroldA** [comment]
Here at ezCompany, we usually only support the last two major versions of a browser, and all older ones should get a usable, but gracefully degraded experience; unless the client is willing to pay for the support for obsolete browsers.

When we ask if they want to pay more for the ~5% (gets even less every month) of the visitors, they often decide to drop full support for <= IE8/9 and go for the graceful option.

**Comment by klonos** [comment]
I think it's time. My quick search for stats showed various numbers that do not go above 2% and it's not limited to a single website.

Related #1362 and #801

**Comment by quicksketch** [comment]
I agree, I think it's time too. Maybe we should take a PMC vote? This is sort of a big deal as far as compatibility goes and is central to our backwards-compatibility philosophy.

Let's slate this for 1.6.0 (Jan 15, 2017) for officially dropping support, rather than last-minute before 1.5.0.

**Comment by klonos** [comment]
Title change because eventually it is going to happen. We just need to decide when 😄 

> This is sort of a big deal as far as compatibility goes and is central to our backwards-compatibility philosophy.

We could mitigate the effect of dropping support by allowing easy detection and reaction for those that do want to keep supporting it (see my previous proposal about getting Modernizr in core). But mind you, according to recent stats, we are talking about 2% or less. So, if one plays the "Backwards compatibility is important" card, we can in response play the "Write code for the majority" + the "Include features for the majority" cards 😄

**Comment by Dinornis** [comment]
I think it is time to drop support for IE8.
Can someone provide any examples of client needing/asking for IE8 compatibility and what is the rationale?

Here are some points for dropping it:
* Support for all versions of IE on Windows XP ended effective April 8, 2014 due to its end-of-life.
* As of January 12, 2016, IE8 is no longer supported on any client or server version of Windows.
* On Windows Vista and Windows Server 2008 only IE9 is supported.
* IE8 does not support significant number of W3C standards.
* IE8 fails the [Acid3](https://en.wikipedia.org/wiki/Acid3) test with a score of 20/100.

**Comment by klonos** [comment]
> - As of January 12, 2016, IE8 is no longer supported on any client or server version of Windows.

A good coincidence that we are releasing 1.6.0 on Jan 15 then 😉 

Copying my own comment from the PMC issue:

> I was about to say that we could have this be done gradually: move the IE8 support code to a core module and have it default to disabled for new installations after say 1.6.0 or 1.7.0, then move that to contrib for people that still need it after 1.7.x

How does that sound for a plan?

**Comment by Dinornis** [comment]
Is that more work than just removing it?

**Comment by jlfranklin** [comment]
Wow.  IE8 is almost more popular than IE9 and IE10 combined.

**Comment by jenlampton** [comment]
> As of January 12, 2016

@klonos we are releasing 1.6 in 2017.

**Comment by MrHaroldA** [comment]
The 'Other' browser is way more popular. Let's support that instead! ;)

If 99% of the visitors benefits by dropping support for a browser which isn't supported anyway; I'd say go for it!

**Comment by klonos** [comment]
> >As of January 12, 2016

>@klonos we are releasing 1.6 in 2017.

@jenlampton yes, I meant that it's past a full year now since official support was dropped. So, we have a(nother) legitimate excuse and we can claim that we have given people enough time. We could move any shim/related code/css hacks to a separate module and have that disabled by default on new installations after 1.6.0. This way we do not mess with existing websites. That then gives us enough time for an official renouncement that we are planning to move it to contrib for 1.7.0.

**Comment by jenlampton** [comment]
Ah okay :)

> This way we do not mess with existing websites. 

Moving code to a new module still messes with existing websites. It would require that people go and turn on the new module after the update. I'm also not comfortable removing anything from core mid-cycle. That should probably wait until Backdrop 2.0.

I think what we should do in 1.6 is **announce** that we are dropping support for IE 8.  This will then  allow us to write cleaner CSS going forward, stop testing our work on IE8, and make core developers happy.

Removing things (or even moving things around) seems both unnecessary and potentially harmful to existing sites. What's the harm in leaving things as they are? We can set a precedent now, and clean up code in 2.0.

**Comment by klonos** [comment]
> Moving code to a new module still messes with existing websites.

True that.

> It would require that people go and turn on the new module after the update.

Don't think do. The module will be disabled by default *only on **new** installations* - not on existing sites (if I got this right, we'd simply have to omit it from the list of `dependencies[]` in the .info files of our core profiles).

> I'm also not comfortable removing anything from core mid-cycle. That should probably wait until Backdrop 2.0. ...Removing things (or even moving things around) seems both unnecessary and potentially harmful to existing sites. What's the harm in leaving things as they are?

👍

**Comment by jenlampton** [comment]
Awesome, pushing this to 2.0 and updating title.

**Comment by Dinornis** [comment]
Starting with version 4.8 WordPress is officially ending support for Internet Explorer versions 8, 9, and 10 (see https://make.wordpress.org/core/2017/04/23/target-browser-coverage/)

**Comment by klonos** [comment]
...one year+ later, it seems that it might make sense to revisit this now. It feels that we could either do this in the 1.x cycle, or just make the call and have the next version of Backdrop be 2.0(?!).

It seems that it might make sense to drop IE9 and below actually(?). Here's stats as shown by https://caniuse.com:

![screen shot 2018-05-14 at 9 18 06 am](https://user-images.githubusercontent.com/2423362/39972725-811ab222-5758-11e8-88a4-e9533e819b3e.png)

![screen shot 2018-05-14 at 9 18 15 am](https://user-images.githubusercontent.com/2423362/39972727-873b67d2-5758-11e8-94b1-7a6c7a002216.png)

There were options for Asia/Europe/Africa/North and South America, which would have helped to make a more informative decision, but there seemed to be errors when I was trying to include stats on a continent level. For now, I hope that the global stats suffice.

Bottom line, we always discuss about the 80/20 use case, but this is not even 0.2% now! Especially if this is to unblock other tasks like #3106, from which we are to gain in performance, security, and features.

**Comment by klonos** [comment]
...by the way, seems we could also use a section about supported (or at least "test-with") browser versions. I was expecting something like that to be available in https://backdropcms.org/requirements

@jenlampton ^^

**Comment by klonos** [comment]
Here's some latest stats from around the net:

https://www.w3counter.com/globalstats.php shows a ~2.5% global usage for IE11, with other versions of IE not even in the top 10 web browsers (September 2018). The breakdown is available at https://www.w3counter.com/trends:

![screen shot 2018-11-01 at 9 00 33 am](https://user-images.githubusercontent.com/2423362/47821398-aeb8d280-ddb4-11e8-8429-e877f8bca504.png)

https://www.w3schools.com/browsers shows a 4% combined for Edge/IE. The breakdown in https://www.w3schools.com/browsers/browsers_explorer.asp shows 1.6% for IE11 and 0.2% for "Older IE".

https://www.statista.com/statistics/544400/market-share-of-internet-browsers-desktop/ shows a ~7% for IE in general (no specific version - August 2018).

http://gs.statcounter.com/browser-market-share shows ~3% for IE in general (no specific version - September 2018).

https://clicky.com/marketshare/global/web-browsers/internet-explorer shows IE11 fluctuating between 2%-6% and "other/Unknown" steady at 0% since August 2018:

![screen shot 2018-11-01 at 8 59 04 am](https://user-images.githubusercontent.com/2423362/47821339-74e7cc00-ddb4-11e8-9471-6fa65293705c.png)

**Comment by klonos** [comment]
Cross-posting from https://github.com/backdrop/backdrop-issues/issues/364#issuecomment-451719891:

The combined, global usage of IE versions 5 through 10 is less than ~0.4% (https://caniuse.com/usage-table). Actually all IE versions + all Edge versions seem to be around ~7% (https://www.w3counter.com/globalstats.php). ...although to be fair, it seems that usage is more around 10%-11% if you limit the scope to desktop browsers alone (https://netmarketshare.com/browser-market-share.aspx). Roughly speaking, everyone seems to be using either Chrome, or Safari (http://gs.statcounter.com) these days.

Time for Backdrop 2.0?? 😄

**Comment by klonos** [comment]
![image](https://user-images.githubusercontent.com/2423362/53710665-8809c600-3e92-11e9-8dce-b2a5dafc37c2.png)

**Comment by klonos** [comment]
https://blogs.windows.com/windowsexperience/2018/12/06/microsoft-edge-making-the-web-better-through-more-open-source-collaboration/

> DECEMBER 6, 2018 9:00 AM
> **Microsoft Edge: Making the web better through more open source collaboration**
>
> ...
> Today we’re announcing that we intend to adopt the Chromium open source project in the development of Microsoft Edge on the desktop to create better web compatibility for our customers and less fragmentation of the web for all web developers.

One thing less to worry about 😉

**Comment by klonos** [comment]
Time far a metrics update:

- https://gs.statcounter.com shows IE steadily much less than 1% worldwide after Dec 2020 (with Safari 18-20% and Chrome 63-65% dominating the market)
- same as above, https://www.w3schools.com/browsers/browsers_explorer.asp shows older versions of IE starting at 0.7% at the start of 2021, and dropping all the way to 0.3% till Aug 2021
- in https://www.w3counter.com/globalstats.php IE is not in the top 10 of browsers used worldwide

I know that we've decided that this is a 2.x task, but perhaps we should reconsider(?)

**Comment by jenlampton** [comment]
I think we can safely add new things without IE support now (see the icon
font in core issue re image masks) but I'm still nervous a about removing
old things.

**Comment by indigoxela** [comment]
IMO it's safe to remove all workarounds specifically done for IE**8** (note the issue title).

Regarding IE11 we probably have to wait a little longer. It's just a matter of time until it's gone, too, now that Microsoft retired it and drops support it their own projects.

But in another discussion it has been mentioned that IE11 is still needed for some screen reader software not supporting other browsers yet. Also a matter of time, I suppose, but dropping might not be appropriate yet.

**Comment by BWPanda** [comment]
@quicksketch mentioned this issue in today's meeting. Bringing this up again for discussion.

**Comment by quicksketch** [comment]
The prompting of this was the official end of life for Internet Explorer (including version 11) as of June 15, 2022, as covered here: https://blogs.windows.com/windowsexperience/2022/06/15/internet-explorer-11-has-retired-and-is-officially-out-of-support-what-you-need-to-know/

Effectively, if your OS is capable of running their newer browser Edge, Microsoft stop opening IE11 and instead open Edge in IE-compatibility mode, which will have support through 2029. I found that the official end-of-life for IE isn't exactly clear-cut though. It's still technically supported in extended-support versions of Windows, going all the way back to Windows 7, see https://docs.microsoft.com/en-us/lifecycle/faq/internet-explorer-microsoft-edge

However, if you're running a normal supported version of Windows (10 or higher), IE basically no longer opens.

**Comment by quicksketch** [comment]
Taking a quick look at IE-specific code we have, the following seem like candidates for removal:

* `html5shiv` library: For rendering HTML5 elements in IE 6-9
* The `#browsers` renderable property (used by `html5shiv`), which lets CSS be added in IE 6-9 specifically for those browsers. No other browsers support conditional comments and IE stopped supporting it in IE 10: https://en.wikipedia.org/wiki/Conditional_comment Looks like the primary code for this is in `backdrop_pre_render_conditional_comments()`.

There is also some XSS protection in `filter_xss()` that is relevant apparently to IE6 only, but as it's part of a giant suite of XSS protections and leaving it has a low impact but removing it could be dangerous, I'm inclined to leave it in place.

**Comment by klonos** [comment]
Once this task here is done (referring to the removal of IE-specific code), we should discuss re policy around IE support in general.

Statistics suggest that global usage is below ~0.5% for all IE versions, and even Edge is struggling below 5%. Most people seem to be using Chrome or Safari to access the internet, and mostly via their mobile devices instead of desktop.

**Comment by quicksketch** [comment]
I filed a PR for removing conditional comments and the HTML5 shim/shiv at https://github.com/backdrop/backdrop/pull/4569.

I did not get into remove various IE-specific CSS entries, as that could require a lot more QA in ensuring the CSS isn't also affecting other browsers in some way.

**Comment by indigoxela** [comment]
Looks great!

The change sure has no impact on display anymore, but wait... can it have? I wonder, how this impacts projects (modules or themes), which still use those browser switch for conditional comments.

For example the [css_injector](https://github.com/backdrop-contrib/css_injector) still provides a "browser switch" for rules.
But if they're gone - does that CSS get applied to all browsers?

Couldn't test quickly, as the Tugboat sandboxes still prevent module installs.

**Comment by indigoxela** [comment]
Did a local test now:

- Fresh install, also installed the css_injector module
- Added a rule: `body {color: red}` with a browser rule "Media: IE8"
- Checked markup - it's wrapped in conditional comments
- Applied patch and flushed caches
- On the home page the text now shows in red :disappointed: 
- Checked markup, the CSS file's no longer wrapped in conditional comments, so the style applies in all browsers

Not sure, what to do with that finding.

It is pretty unlikely that someone _recently_ used the "browser" switch for css or any other head tag. But in older sites (older themes) they still might be in use. And that could lead to visual glitches.

I didn't test any other head tag (like js).

**Comment by quicksketch** [comment]
Thanks @indigoxela, good points. I think we should check if `browsers` is specified and just silently not output the CSS file at all. That will prevent already existing themes or modules from accidentally leaking their IE-only CSS to other browsers.

**Comment by kiamlaluno** [comment]
`browsers` is also used to add CSS style to browsers that are not Internet Explorer. In that case, the CSS styles should be added, I guess.

**Comment by quicksketch** [comment]
> `browsers` is also used to add CSS style to browsers that are not Internet Explorer.

I don't think this is the case. Conditional comments [only ever worked on IE 4 through 10](https://en.wikipedia.org/wiki/Conditional_comment). No other browser has ever supported them. I think at the time they were implemented in Drupal, it was conceivable that other browsers might be support conditional comments in the future, but that never happened. I think that if any other browser was specified in `browsers`, it already wasn't usable CSS, but I'll need to double-check that.

**Comment by kiamlaluno** [comment]
[`backdrop_pre_render_conditional_comments()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21common.inc/function/backdrop_pre_render_conditional_comments/1) accepts either `'!IE' => TRUE` which renders the element in all the browsers but Internet Explorer, or `'!IE' => FALSE` which does not render the element when the browser is not Internet Explorer.

I tried with the following code.

```php
$elements = array(
  '#browsers' => array(
    'IE' => FALSE,
    '!IE' => TRUE,
  ),
);

var_dump(backdrop_pre_render_conditional_comments($elements));
```

The output has been the following. (I used `\n` where a new-line character was printed.)

```none
array(3) {
  ["#browsers"] =>
  array(2) {
    ["IE"] => bool(false)
    ["!IE"] => bool(true)
  }
  ["#prefix"] => string(20) "\n<!--[if !IE]><!-->\n"
  ["#suffix"] => string(17) "<!--<![endif]-->\n"
}
```

**Comment by quicksketch** [comment]
Thanks @kiamlaluno. I think you're right. I'll have to think on this a bit more.

**Comment by indigoxela** [comment]
@kiamlaluno many thanks for pointing this out. :+1: 

Although this will hardly ever be in use, it's still possible and needs some consideration when it comes to backwards compatibility.

**Comment by quicksketch** [comment]
I decided to try for a smaller scope and just remove the html5shiv for now. It's a much smaller step but it will be nice to remove that library from being output on every single page of Backdrop output. PR at https://github.com/backdrop/backdrop/pull/4611 needs review.

**Comment by indigoxela** [comment]
Much better! We're still getting rid of this old `if lte IE 8`, but it doesn't break anything.

How about adding a hint to documentation? Currently nobody knows that this is outdated stuff and a candidate for removal, "browsers" still gets advertised in the docblock of backdrop_add_js() and backdrop_add_css(). No idea about appropriate wording.

**Comment by quicksketch** [comment]
Thanks @indigoxela. I separated out deprecating `browsers` into a separate PR at https://github.com/backdrop/backdrop/pull/4615. This approach now leaves the `browsers` option in place but documents it as deprecated, and provides a fairly safe behavior for sites that may have used it in the past.

Currently, to target IE-only, the API also requires you specify how other browsers should be treated. For example `array('IE' => 'lt IE 7', '!IE' => FALSE)` would target IE6 and lower only. The new approach only pays attention to the `'!IE'` (other browsers) value. If it is `TRUE`, then the file is output completely normally. If it is `FALSE`, the file is not output at all. Anything passed into the `IE` value is completely ignored.

I think maybe we could merge the removal of the IE shiv (https://github.com/backdrop/backdrop/pull/4611) and continue to work on conditional comments removal/deprecation.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/4611 to get started. We can continue discussing removing conditional comments here.

**Comment by quicksketch** [comment]
Adding the 1.27.0 milestone since removing the IE shiv is now a part of the next release.

**Comment by indigoxela** [comment]
This seems to work properly now. :+1: 

I wonder if we could also remove [these default values](https://github.com/backdrop/backdrop/blob/1.x/core/includes/common.inc#L3514). Probably not, as this has influence on grouping, but...

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/4615 to clear this long-standing issue off our plate. Now with both `#browsers` deprecated and no longer having any effect, combined with the HTML 5 shiv removed, I think this issue is sufficiently complete. We can still remove further IE-specific code (such as our IE support in Bootstrap, see #6345) in separate issues.

**Comment by quicksketch** [comment]
Reopening until we have a change record for the new `#browsers` behavior created.

**Comment by quicksketch** [comment]
I completed the change record and published it at https://docs.backdropcms.org/change-records/internet-explorer-conditional-comments-deprecated-browsers-property.

**Comment by quicksketch** [comment]
> other old browser specific identifiers that have been removed

The `browsers` property could only ever target specific IE versions. They can't target any other browsers other than a general "not IE" target. So the behavior we have now is that all IE-version targeting no longer does anything; and if a `browsers` property specified "not IE", then we just output the CSS/JS normally without any conditional comment at all.

Is that what you mean by other browser identifiers?
